src/utp_stream.cpp

```diff
@@ -2992,6 +3024,10 @@ bool utp_socket_impl::incoming_packet(span<char const> b
 			// After that has happened we know the remote side has all our
 			// data, and we can gracefully shut down.
 
+			// we should still ack any incoming data to prevent potential
+			// timeouts/resends at the other end
+			if (ph->get_type() == ST_DATA) defer_ack();
```

Comment:
I imagine this is the key change for fixing the timeouts in the tests, right? since those tests never cause the socket to stall

Reply:
Deferring acks here is so we can continue to ack data packets when we're the first end to send the fin packet.

I think the key change for fixing the timeouts in the tests is in `maybe_inc_acked_seq_nr()`. From the test's utp log, prior to this PR we had this:
```
[000004732088] 0x562466e0ca60: incoming packet seq_nr:39341 ack_nr:54427 type:ST_STATE id:14063 size:0 timestampdiff:32623 timestamp:5199507 our ack_nr:39341 our seq_nr:54427 our acked_seq_nr:54426 our state:FIN_SENT
```
Note that the incoming packet is trying to ack the fin packet (54427), however our acked_seq_nr hasn't been increased.

`maybe_inc_acked_seq_nr()` now allows `acked_seq_nr` to reach `seq_nr` once we're in the fin-sent state. After this PR, for the same packet we now get the following. Note that our acked_seq_nr has now increased to where it should be.
```
[000004732088] 0x558860f04c20: incoming packet seq_nr:39341 ack_nr:54427 type:ST_STATE id:14063 size:0 timestampdiff:32623 timestamp:5199507 our ack_nr:39341 our seq_nr:54427 our acked_seq_nr:54427 our state:FIN_SENT
[000004732088] 0x558860f04c20: FIN acked

```
For `maybe_inc_acked_seq_nr()`, I think we also need to have the fin packet stored in `m_outbuf` to stop it being prematurely ack'd, which we also now do. The tests still timeout without both of these changes.